windev.ps1

```diff
@@ -1,55 +1,170 @@
+    # Match the entire 'Version:' header and extract the script's version number directly using RegEx capture groups.
+    $RemoteWinUtilVersion = (([regex]($VersionExtractionRegEx)).Match($RawWinUtilScript).Groups[2].Value)
+    $LocalWinUtilVersion = (([regex]($VersionExtractionRegEx)).Match((Get-Content $WinUtilScriptPath)).Groups[2].Value)
+
+    # Check if WinUtil needs an update and either download a fresh copy or notify the user its already up-to-date.
+    if ([version]$RemoteWinUtilVersion -ne [version]$LocalWinUtilVersion) {
+        Write-Host "WinUtil is out-of-date. Downloading WinUtil '$($RemoteWinUtilVersion)'..." -ForegroundColor DarkYellow
+        Invoke-RestMethod $WinUtilReleaseURL -OutFile $WinUtilScriptPath
+    } else {
+        Write-Host "WinUtil is already up-to-date. Skipped update checking." -ForegroundColor Yellow
+    }
+}
+
+# Function to start the latest release of WinUtil that was previously downloaded and saved to '$env:TEMP\winutil.ps1'.
+function Start-LatestWinUtil {
+    param (
+        [Parameter(Mandatory = $false)]
+        [array]$WinUtilArgumentsList
+    )
+
+    # Setup the commands used to launch WinUtil using Windows Terminal or Windows PowerShell/PowerShell Core.
+    $PowerShellCommand = if (Get-Command pwsh.exe -ErrorAction SilentlyContinue) { "pwsh.exe" } else { "powershell.exe" }
+    $ProcessCommand = if (Get-Command wt.exe -ErrorAction SilentlyContinue) { "wt.exe" } else { $PowerShellCommand }
+
+    # Setup the script's launch arguments based on the presence of Windows Terminal or Windows PowerShell/PowerShell Core:
+    # 1. Windows Terminal needs the name of the process to start ($PowerShellCommand) in addition to the launch arguments.
+    # 2. Windows PowerShell and PowerShell Core can receive and use the launch arguments as is without extra modification.
+    $WinUtilLaunchArguments = "-ExecutionPolicy Bypass -NoProfile -File `"$WinUtilScriptPath`""
+    if ($ProcessCommand -ne $PowerShellCommand) {
+        $WinUtilLaunchArguments = "$PowerShellCommand $WinUtilLaunchArguments"
+    }
+
+    # If WinUtil's launch arguments are provided, append them to the end of the list of current launch arguments.
+    if ($WinUtilArgumentsList) {
+        $WinUtilLaunchArguments += " " + $($WinUtilArgumentsList -join " ")
+    }
+
+    # If the WinUtil script is not running as administrator, relaunch the script with administrator permissions.
+    if (!$ProcessIsElevated) {
+        Write-Host "WinUtil is not running as administrator. Relaunching..." -ForegroundColor DarkCyan
+        Write-Host "Running the selected WinUtil release: Version '$($WinUtilReleaseTag)'." -ForegroundColor Green
+        Start-Process $ProcessCommand -ArgumentList $WinUtilLaunchArguments -Wait -Verb RunAs
+    } else {
+        Write-Host "Running the selected WinUtil release: Version '$($WinUtilReleaseTag)'." -ForegroundColor Green
+        Start-Process $ProcessCommand -ArgumentList $WinUtilLaunchArguments -Wait
     }
```

Comment:
Maybe `Write-Host "Running the selected WinUtil release: Version '$($WinUtilReleaseTag)'." -ForegroundColor Green` can be moved before the if, as it is repeated.
Also, i'm curious if there is any concern stopping us from running it as administrator either way? Could something like below be a problem? (I'll also leave this up to you either way)
```suggestion
    Write-Host "Running the selected WinUtil release: Version '$($WinUtilReleaseTag)'." -ForegroundColor Green
    if (!$ProcessIsElevated) {
        Write-Host "WinUtil is not running as administrator. Relaunching..." -ForegroundColor DarkCyan
    }
    Start-Process $ProcessCommand -ArgumentList $WinUtilLaunchArguments -Wait -Verb RunAs
```

Reply:
I wouldn't think so, but I don't think it'd be a good idea to use `-Verb RunAs` unconditionally, it may save a few bytes but it could also open up security issues, for example: If an attacker were to compromise this repository they could replace `winutil.ps1` with a malicious script that stole all of your personal information, emails, passwords, etcetera.
As for the log message you mentioned, I'll try to move it and see if it appears in the same place.

Reply:
@momcilovicluka Unfortunately, I couldn't find an easy way to move the log message in a way that retains the format, if you have a suggestion for a color I can use to distinguish between admin and non-admin that is easily recognizable I can implement that instead. Would that work?

Reply:
Yeah, i agree about not using it unconditionally, i was just curious. About it being security issue, it doesn't change anything? In my example it always tries to run it as admin whether it was started as admin or not, but in original, it would first check if it was not admin to start it as admin, and if it was, it is again being started with admin privileges either way. It is effectively doing the same thing. It always is started as admin since the script doesn't work without that. But if the attacker compromises repository, then we have bigger issues 🙃

Also is there any way to make it execute in the same window? When i tested it from main it executed everything in the same window, but here it creates a new window always. I have tried with `-NoNewWindow`, but that doesn't seem to work.

Reply:
Sadly, no, I tried to do that but the way `-Verb RunAs` works is that it opens a new window when running the script, there's no easy way around that. `-PassThru` can work but then you'd have to redirect the output of the script and that... Would probably require all of WinUtil to be rewritten.

Reply:
As for the current version of `windev.ps1`, when running it as a non-elevated user the following error occurs:
```
Unable to determine latest pre-release version.
Using latest Full Release
Winutil needs to be run as Administrator. Attempting to relaunch.
Start-Process:
Line |
  68 |      Start-Process $processCmd -ArgumentList "$powershellcmd -Executio …
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | This command cannot be run due to the error: An error occurred trying to start process 'C:\Users\#####\AppData\Local\Microsoft\WindowsApps\wt.exe' with working directory 'C:\Users\#####'. The parameter is incorrect.
Winutil needs to be run as Administrator. Attempting to relaunch.
```

Reply:
@momcilovicluka Also, I implemented your suggestion and moved the log message outside of the `if (!$ProcessIsElevated)` statement, I tried a few more things but the colors looked really similar to each other.